core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1178,7 +1185,63 @@ private StageExecution getStageExecution(StageId stageId)
         private static List<SubPlan> sortPlanInTopologicalOrder(SubPlan subPlan)
         {
             ImmutableList.Builder<SubPlan> result = ImmutableList.builder();
-            Traverser.forTree(SubPlan::getChildren).depthFirstPostOrder(subPlan).forEach(result::add);
+            // This prioritizes scheduling the build side of joins ahead of the probe side,
+            // because joins cannot progress until the build side is complete.
+            SuccessorsFunction<SubPlan> getChildren = node -> {
+                Map<PlanFragmentId, SubPlan> sourceSubPlans = node.getChildren().stream()
+                        .collect(toMap(plan -> plan.getFragment().getId(), plan -> plan));
+                ImmutableList.Builder<SubPlan> children = ImmutableList.builder();
+                node.getFragment().getRoot().accept(new SimplePlanVisitor<Void>()
+                {
+                    @Override
+                    public Void visitJoin(JoinNode node, Void context)
+                    {
+                        node.getRight().accept(this, context);
+                        node.getLeft().accept(this, context);
+                        return null;
+                    }
+
+                    @Override
+                    public Void visitSemiJoin(SemiJoinNode node, Void context)
+                    {
+                        node.getFilteringSource().accept(this, context);
+                        node.getSource().accept(this, context);
+                        return null;
+                    }
+
+                    @Override
+                    public Void visitSpatialJoin(SpatialJoinNode node, Void context)
+                    {
+                        node.getRight().accept(this, context);
+                        node.getLeft().accept(this, context);
+                        return null;
+                    }
+
+                    @Override
+                    public Void visitIndexJoin(IndexJoinNode node, Void context)
+                    {
+                        node.getIndexSource().accept(this, context);
+                        node.getProbeSource().accept(this, context);
+                        return null;
+                    }
+
+                    @Override
+                    public Void visitRemoteSource(RemoteSourceNode node, Void context)
+                    {
+                        for (PlanFragmentId fragmentId : node.getSourceFragmentIds()) {
+                            SubPlan child = sourceSubPlans.remove(fragmentId);
+                            requireNonNull(child, "PlanFragmentId %s does not appear in children of %s".formatted(fragmentId, node));
```

Comment:
Nit:
```suggestion
                            requireNonNull(child, "PlanFragmentId %s does not appear in sources of %s".formatted(fragmentId, node));
```

Reply:
oh boy, i can take a look at unit tests, but that sounds difficult.

I completely agree about benchmarks - i would love to learn how to test a local build of trino in a cluster. Maybe you can help me out with that?